walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -39,46 +41,90 @@
+                }else if(availableTimes.get(i).contains(availableOptionalTimes.get(j))){
+                    slots.add(TimeRange.fromStartEnd(availableOptionalTimes.get(j).start(), availableOptionalTimes.get(j).end(),false));
+                }else if(availableOptionalTimes.get(j).contains(availableTimes.get(i))){
+                    slots.add(TimeRange.fromStartEnd(availableTimes.get(i).start(), availableTimes.get(i).end(),false));
+                }
             }
         }
 
-        //Get everthing in between the booked time slots as eligible slots for the request
-        long duration = request.getDuration();
-        
-        if((bookedTimes.get(0).start() - TimeRange.START_OF_DAY) >= duration){
-            slots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, bookedTimes.get(0).start(), false));
+        // If no available overlap is available, return the available times for required attendees.
+        if(slots.size() == 0 && availableOptionalTimes.size() != 0){
+            return availableTimes;
         }
 
-        for(i = 0; i < bookedTimes.size() - 1; i++){
-            if((bookedTimes.get(i+1).start() - bookedTimes.get(i).end()) >= duration){
-                slots.add(TimeRange.fromStartEnd(bookedTimes.get(i).end(), bookedTimes.get(i+1).start(), false));
-            }
+    }
+ 
+    return slots;
+  }
+ 
+  public List<TimeRange> getAvailableSlots(List<TimeRange> timesArray, long duration){
+    
+    List<TimeRange> returnSlots = new ArrayList<>();
+ 
+    if(timesArray.size() == 0){
+        returnSlots.add(TimeRange.WHOLE_DAY);
+        return returnSlots;
+    }
+ 
+    //Sort the booked times by start of day
+    Collections.sort(timesArray, TimeRange.ORDER_BY_START);
+ 
+    //Merge overlapping or nested timeslots into a single chunk of time
+    int i = 0;
+    while(i < timesArray.size()-1){
+        if(timesArray.get(i).overlaps(timesArray.get(i+1)) || timesArray.get(i).contains(timesArray.get(i+1))){
+            int endTime = Math.max(timesArray.get(i).end(), timesArray.get(i+1).end());
+            timesArray.set(i, TimeRange.fromStartEnd(timesArray.get(i).start(), endTime, false));
+            timesArray.remove(i+1); 
+        }else{
+            i++;
         }
-        
-        if((TimeRange.END_OF_DAY - bookedTimes.get(bookedTimes.size() - 1).end()) >= duration){
-            slots.add(TimeRange.fromStartEnd(bookedTimes.get(bookedTimes.size() - 1).end(), TimeRange.END_OF_DAY, true));
+    }
+ 
+    //Get everthing in between the booked time slots as eligible slots for the request.
+    if((timesArray.get(0).start() - TimeRange.START_OF_DAY) >= duration){
+        returnSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, timesArray.get(0).start(), false));
+    }
+ 
+    for(i = 0; i < timesArray.size() - 1; i++){
+        if((timesArray.get(i+1).start() - timesArray.get(i).end()) >= duration){
```

Comment:
instead of doing a duration check at 3 different places, you could instead just combine all the time slots and then perform 1 final filter based on duration at the end. That is better for code maintenance and readability.

Reply:
I agree. Will implement

Reply:
Resolved